analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs

```diff
@@ -35,7 +35,6 @@ self switch
             {
                 { TypeKind: TypeKind.Struct } => true,
                 ITypeParameterSymbol { IsValueType: true } => true,
-                ITypeParameterSymbol { HasReferenceTypeConstraint: false, ConstraintTypes: { IsEmpty: false } constraintTypes } => constraintTypes.Any(x => x.SpecialType == SpecialType.System_Enum),
```

Comment:
I removed this line from the `IsStruct` method because the Enum constraint is a beast. If you have `where T: Enum` it basically means T is constraint to anything deriving from Enum, which means all types declared by `enum Bla { }`. So all `T` are value types. But `T?` is not treated the same as when `where T: struct`:
Two methods with a parameter `T? t` and called like so `M<Base64FormattingOptions>(Base64FormattingOptions.InsertLineBreaks);` behave differently depending on the constraint for `T`:

| Call                | where T: struct, Enum                             | where T: Enum, new()           |
|---------------------|---------------------------------------------------|--------------------------------|
|      t.HasValue     |                        True                       |        Does not compile        |
| t                   | InsertLineBreaks                                  | InsertLineBreaks               |
| t.GetType()         | System.Base64FormattingOptions                    | System.Base64FormattingOptions |
| typeof(T)           | System.Base64FormattingOptions                    | System.Base64FormattingOptions |
| typeof(T?)          | System.Nullable`1[System.Base64FormattingOptions] | System.Base64FormattingOptions |
| GetUnderlyingType() | System.Base64FormattingOptions                    |                                |
| default(T)          | None                                              | None                           |
| default(T?)         |                                                   | None                           |
| default T           | None                                              | None                           |
| default T?          |                                                   | None                           |
| new T()             | None                                              | None                           |
| new T?()            |                                                   | None                           |

See [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwUUCyVAPAEIQDOApgGwAsAYgHtkkBAlgBzAPIAHMQJhsAfAApWnXoOERREmXIUA6AJIKOyBABlYHZsg4QA1mwCUAbnpM13fkJFiYUrJQ8krKGBwAZhCoADYIbvSELOzemn66QSEqXhq+2v6B+mzGpuZWMDZ2ji7u5HRJOT5aOgF6wQoq4VGx8bUU+DwABAyMACoqowD8g/GDAO4AFmYcg6Mgg2wIaADGCHCDAKIwqGAUAN4Ug1eD+FQAnMoJ5Nc398oARIvLq+ubO3uHY5gd5PF63B4AEneCAMAAl2AA1CAxVArF4vdZnGHwthIlEcAC+INqYLeUIQ6MpVMxCCJoOu4OU5IMAHEOAhRgBPaQcR5UwaY5QUgC8wsGxxiMUG03eAHUBLEMDMFsgBHN3usYWyOdzec5nHSSQyydDdQIIspRs5+VcaWaLVbDZdjZDTTzzZbJtb+Xb3Q6vU7ni6me9tQBVGDhZAxTkSLk8x6YgBysRiEGAMQ4rPZEajMbjuqF9s9+sDpNdXWicUt3upgzOlZ6NbLwahjerU1r6Mx7YQJZbV2dV1GgyVYt7RuH0wwMEG48iVYQk9eFYXPVWNoF9YwA5XId7q2mPu3MF3jKhFTmqz5m63Z0v14NxKHe4vHCvUxvNsxD8/T6eBK0JQQz1GMEzTLMXx2D8gInPsl6POcL6MvSVyMp8SzQWssFgPB76PM+QZoSa2KIsiqLfvW7wACICBwbDigIFLbAIYDSFAmbvGeJG3nWWLca6WrsvGerHli2Y6gm/7LuebocB6Vqbr68kOtJyEkcWnaUVimkBoR5YhuGkZmPmAQiYm9YppK6aZhJuYmbGZmFggumlvprbvL2NbaV5jrucRq7dB2Xpib5eluC8L4jmOo5rnEy5TKOs7zkFS7qYFi4bj5An7nFFKJaFp7+a+7y/l+2llWpRGHjMyXivhqElb+kzlXxCBFQBFBAA) for a complete example.

My bottom line is that the `Enum` constraint does not imply `struct` and as such `T?` behaves like an unconstraint or class-constraint type parameter. Therefore I moved the Enum test to it's own dedicated method.
As a result, in a case like `(T t) where T: Enum` `t` is no longer getting the `ObjectConstrint.NotNull` applied at the moment (see the change [here](https://github.com/SonarSource/sonar-dotnet/pull/6949/files#diff-958483de1a8a6b3e2ff9596211ccadad0647164c02e091f40864b7f998e2705fR566)). We may add the `IsEnum` check from below to the `NonNullableValueTypeCheck` to circumvent this problem but I wanted to discuss this problem before that.

I would further suggest to add a new rule idea: "Don't apply the Enum constraint without specifying the `struct` constraint as well." The reasoning is given in the table above: `T?` is just not behaving as it should.